File: backend/src/llm/providers/anthropic-provider.ts
import Anthropic from '@anthropic-ai/sdk';
import { ModelProvider, ModelConfig, ModelResponse, StreamChunk } from '../model-provider';

export class AnthropicProvider extends ModelProvider {
  private client: Anthropic;
  
  constructor(config: ModelConfig) {
    super(config);
  }
  
  async initialize(): Promise<void> {
    this.client = new Anthropic({
      apiKey: this.config.apiKey || process.env.ANTHROPIC_API_KEY
    });
  }
  
  async generate(
    prompt: string,
    options?: Partial<ModelConfig>
  ): Promise<ModelResponse> {
    this.validateInput(prompt);
    const config = this.mergeConfig(options);
    
    const startTime = Date.now();
    
    try {
      const response = await this.client.messages.create({
        model: config.name,
        max_tokens: config.maxTokens,
        temperature: config.temperature,
        top_p: config.topP,
        messages: [{ role: 'user', content: prompt }],
        stop_sequences: config.stopSequences.length > 0 ? config.stopSequences : undefined
      });
      
      const content = response.content[0];
      const text = content.type === 'text' ? content.text : '';
      
      return {
        text,
        tokensUsed: response.usage.input_tokens + response.usage.output_tokens,
        finishReason: response.stop_reason || 'stop',
        metadata: {
          model: response.model,
          latency: Date.now() - startTime,
          inputTokens: response.usage.input_tokens,
          outputTokens: response.usage.output_tokens,
          stopSequence: response.stop_sequence
        },
        cost: this.getCostEstimate(
          response.usage.input_tokens,
          response.usage.output_tokens
        )
      };
      
    } catch (error: any) {
      throw new Error(`Anthropic API error: ${error.message}`);
    }
  }
  
  async *streamGenerate(
    prompt: string,
    options?: Partial<ModelConfig>
  ): AsyncIterableIterator<StreamChunk> {
    this.validateInput(prompt);
    const config = this.mergeConfig(options);
    
    try {
      const stream = await this.client.messages.create({
        model: config.name,
        max_tokens: config.maxTokens,
        temperature: config.temperature,
        top_p: config.topP,
        messages: [{ role: 'user', content: prompt }],
        stop_sequences: config.stopSequences.length > 0 ? config.stopSequences : undefined,
        stream: true
      });
      
      for await (const chunk of stream) {
        if (chunk.type === 'content_block_delta' && chunk.delta.type === 'text_delta') {
          yield {
            text: chunk.delta.text,
            isComplete: false
          };
        } else if (chunk.type === 'message_stop') {
          yield {
            text: '',
            isComplete: true,
            metadata: {
              finishReason: 'stop'
            }
          };
          break;
        }
      }
      
    } catch (error: any) {
      throw new Error(`Anthropic streaming error: ${error.message}`);
    }
  }
  
  async embed(texts: string[]): Promise<number[][]> {
    // Anthropic doesn't provide embedding API, use a fallback or throw error
    throw new Error('Anthropic does not provide embedding API. Use OpenAI or other providers for embeddings.');
  }
  
  estimateTokens(text: string): number {
    // Anthropic의 토큰 계산 (대략적)
    // Claude는 대략 1 토큰당 3.5-4 문자
    return Math.ceil(text.length / 3.5);
  }
  
  getCostEstimate(inputTokens: number, outputTokens: number): number {
    // Anthropic 가격표 (2024년 기준)
    const pricing: Record<string, { input: number; output: number }> = {
      'claude-3-opus-20240229': { input: 0.000015, output: 0.000075 },
      'claude-3-sonnet-20240229': { input: 0.000003, output: 0.000015 },
      'claude-3-haiku-20240307': { input: 0.00000025, output: 0.00000125 },
      'claude-2.1': { input: 0.000008, output: 0.000024 },
      'claude-2.0': { input: 0.000008, output: 0.000024 }
    };
    
    const modelPricing = pricing[this.config.name] || pricing['claude-3-sonnet-20240229'];
    
    return (inputTokens * modelPricing.input) + (outputTokens * modelPricing.output);
  }
}

File: backend/src/llm/providers/anthropic_provider.py
import os
from typing import Any, AsyncIterator, Dict, List, Optional

import anthropic

from ..base import ModelProvider, ModelResponse


class AnthropicProvider(ModelProvider):
    """Anthropic Claude 프로바이더"""

    async def initialize(self) -> None:
        self.client = anthropic.AsyncAnthropic(api_key=os.getenv("ANTHROPIC_API_KEY"))

    async def generate(
        self, prompt: str, options: Optional[Dict[str, Any]] = None
    ) -> ModelResponse:
        response = await self.client.messages.create(
            model=self.config.name,
            messages=[{"role": "user", "content": prompt}],
            max_tokens=self.config.max_tokens,
            temperature=options.get("temperature", self.config.temperature)
            if options
            else self.config.temperature,
        )

        return ModelResponse(
            text=response.content[0].text,
            tokens_used=response.usage.input_tokens + response.usage.output_tokens,
            finish_reason=response.stop_reason,
            metadata={"model": response.model},
        )

    async def stream_generate(
        self, prompt: str, options: Optional[Dict[str, Any]] = None
    ) -> AsyncIterator[str]:
        async with self.client.messages.stream(
            model=self.config.name,
            messages=[{"role": "user", "content": prompt}],
            max_tokens=self.config.max_tokens,
        ) as stream:
            async for text in stream.text_stream:
                yield text

    async def embed(self, texts: List[str]) -> List[List[float]]:
        # Anthropic doesn't provide embeddings, use a fallback
        raise NotImplementedError("Anthropic doesn't provide embedding models")

    def estimate_tokens(self, text: str) -> int:
        return len(text) // 4

    def get_cost_estimate(self, input_tokens: int, output_tokens: int) -> float:
        # Claude-3 기준 비용
        return (input_tokens * 0.000015) + (output_tokens * 0.000075)
